design-docs/005-manifest-split.md

```diff
@@ -0,0 +1,406 @@
+and the current design must not block it.
+
+We define _manifest sets_. A manifest set is a group of manifests that
+is intended to optimize for certain types of arrays. We let the user define
+the sets, but we offer good defaults that should work reasonably well in
+common cases.
+
+We let the user indicate _rules_ that assign arrays to
+a specific manifest set. These rules can be based on array paths or number
+of chunks. In the future we could add more power to the rules system.
+
+Other new feature we provide is the ability to preload certain manifests.
+These manifests will be fetched in the background as soon as a session is
+created, speeding up interactive use. In the future, we may add the ability
+to prefetch manifests for certain branches/tags as soon as the Repository is
+open.
+
+Manifest sets, array rules, and manifest prefetch, are configured in the
+persistent configuration of the repository and can be overloaded on open as
+any other configuration value.
+
+The system may seem complicated when looking at all the knobs, but:
+
+* Average users won't need to do any tuning, default configuration is good enough
+* When they need to, because their repo is very particular or very large, they can
+* It's not hard to implement (see algorithm below). At least not much harder
+than any other mechanism. The difficulties lie in generating the split
+manifests, and that is needed with any system. The extra work of classifying what arrays
+go where is simple.
+
+To understand how the feature works, we show an example configuration. Comments
+in the yaml file explain what the different settings do.
+
+```yaml
+chunk-manifests:
+  sets:                           # the lists of manifest sets and their properties
+                                  # it's a list because order is important
+
+    - coord1:                     # a name for the set, used later with rules
+
+        max-manifest-size: 10000  # this set will contain manifests with <= 10k references
+
+        arrays-per-manifest: null # pack this many arrays in each manifest
+                                  # max-manifest-size and arrays-per-manifest are
+                                  # mutually exclusive
+
+        overflow-to: coord2       # if an array cannot fit in the set, send it to set coord2
+                                  # if no overflow-to is indicated it goes to
+                                  # the default set (see below)
+
+        cardinality: 2            # this set cannot have more than two manifests
+                                  # cardinality = null means unlimited
+
+    - coord2:
+        max-manifest-size: 10000
+                                  # overflow-to: default       # by default
+                                  # arrays-per-manifest: null  # by default
+                                  # cardinality: 1             # by default
+
+    - big-array:
+        max-manifest-size: null
+        arrays-per-manifest: 1
```

Comment:
I realized that `arrays-per-manifest`  is useful for building up the big ERA5 (or similar) dataset sequentially array-by-array.

Reply:
Yes, that too, but also to minimize latency if you are only reading from a couple of variables, which I suspect is very common.